### ISSUE TITLE
Downgrade to an unmanned mission?

### ISSUE THREAD

**Issue by peyre** [raceintospace/raceintospace#241]
There have been times I've planned a manned orbital, for instance, but when I go to assemble it the hardware isn't quite ready and I'd like to fly an unmanned orbital instead.  That would be a really handy option, but . . . is it realistic?  Can a manned mission be changed to a dummy one at short notice like that, or do dummy missions require substantial preparation _as_ dummy missions (to fine-tune the automation, etc.)?  Does anyone know?  (I wouldn't want to do this if it's not a realistic option.)

**Comment by peyre** [comment]
Incidentally, I asked NASA about this one.  The public can send a question on its Web site.  All I got back was a canned response referring me to NASA's history, so that was a waste of effort.

**Comment by peyre** [comment]
I'v contacted Erik Anderson about this.  He was one of Fritz's main sources of inspiration for the creation of BARIS, and is very knowledgeable about it and the Space Race.  He says it should have been doable in real life, that even more complex missions like docking tests should have been downgradable to unmanned missions, and that this is all allowed by the rules of _Liftoff!_.  So I think we're Go for this one.  I'll work out the details and implement it when I can.

**Comment by peyre** [comment]
I have added downgrade-to-unmanned options in the rush.cpp file.  These are the changes; each full line is as it is now, with the original items below it.
    {03, 00, 00, 00, 00, 00}, // 2
     00
    {02, 05, 03, 00, 00, 00}, // 4
           00, 00
    {04, 02, 15, 00, 00, 00}, // 14
                 00
    {18, 00, 00, 00, 00, 00}, // 16
     00
    {14, 04, 15, 00, 00, 00}, // 20
     04, 00, 00
    {16, 18, 00, 00, 00, 00}, // 21
           00
    {05, 00, 00, 00, 00, 00}, // 42
     00
    {04, 42, 00, 00, 00, 00}, // 43
           00
    {42, 05, 00, 00, 00, 00}, // 45
           00
    {43, 04, 45, 42, 05, 00}, // 46
                 00, 00, 00

**Comment by rnyoakum** [comment]
From what I saw this commit has been reverted, so I'm reopening the issue.

**Comment by peyre** [comment]
Thanks Ryah.  You think it's doable to change the code to make it possible to downgrade to an unmanned mission?

**Comment by peyre** [comment]
Since this has been reopened, it might be helpful (or at least amusing) to show off the screenshots of what happens on such a mission.

![Launchpad](https://user-images.githubusercontent.com/3118014/65784078-26e36180-e106-11e9-83f3-642536c43dba.png)
![Dead](https://user-images.githubusercontent.com/3118014/65784093-2d71d900-e106-11e9-918f-85a62e577a63.png)
![Dead_on_USub](https://user-images.githubusercontent.com/3118014/65784100-31056000-e106-11e9-95c6-886f1de0b2c3.png)

**Comment by rnyoakum** [comment]
It should be possible to get this working. A couple of thoughts:

- This should probably be a configurable option, for those people who feel it's a significant change from the original game and wish to disable this functionality.
- When the primary and backup flight crew for a mission are unavailable (due to a retirement/news event) the mission is scrubbed automatically. If downgrading to an unmanned version is an option, how should that be handled?
- Same for the news event where an astronaut death causes all manned flights to be cancelled. Should they be downgraded automatically, scrubbed, or are we building a new user interface to handle this? If downgraded automatically, how is the replacement mission chosen?

**Comment by peyre** [comment]
- I would have figured people just wouldn't use the option if they wanted to play the game the old way, but then, we haven't worked from that assumption with some other changes we've made.  I have no objection to making that configurable, though it might make the code look a little messy.
- These other two are very good questions.  The easiest, of course, would be to leave it as is and try to get by on the assumption that the suddenness of losing the crew(s) has made it impractical to turn it around as an unmanned mission.  On the other hand it would be handy if it were downgraded automatically so you could at least get a dummy test out of the mission, if you wanted it.  If you didn't want it you could simply scrub the mission.  I think it would be dropped to its unmanned counterpart: suborbital -> U suborbital, orbital -> U orbital, docking -> U docking.  A dialog box asking how you want to handle it would be nice, but a lot of work to build, and my guess it wouldn't be worth the effort.  One consideration if we downgraded automatically is that we'd have to change the newscast.  We can change the text easily enough, but the voice will still say the same thing; 'course we could work around that a bit I guess.

**Comment by rnyoakum** [comment]
This should be working per commit ba8b2cc43456727140d4a79eba8a75f10918198d. The updated downgrade system, like the system it replaces, is a little fragile. It trusts that the table in data/gamedata/downgrades.json is accurate and doesn't contain any missions with incompatible hardware requirements. If that assumption is mistaken, the results are unspecified.

The hardware requirements is a pretty strong requirement - for one, it essentially prevents any two-crew missions from being downgraded. Joint missions & single launch missions cannot be swapped. Don't downgrade anything which has an optional docking module to a mission which requires one - the hardware has already been assigned in the VAB. It's possible to downgrade a mission with a DM to one without - the DM will be assigned until the next turn, but _probably_ won't be consumed; it may also be possible to free it up by reassigning hardware in the VAB.

I'll try to tighten up the hardware issues at some point and make the system more resilient. Currently, there isn't much protection surrounding those issues (Ex: #292) so I think the added functionality - used properly - is worth it.

**Comment by peyre** [comment]
That's great!  I'll test it this morning.  Adding some ability to change the components of a mission is something new.  That opens exciting possibilities, though I can see the can of worms potential here too.  It might be just as well that the downgrade options aren't moddable, that they have to be changed in the code itself.

    On Friday, November 1, 2019, 7:14:09 PM PDT, Ryan Yoakum <notifications@github.com> wrote:  
 
 
This should be working per commit ba8b2cc. The updated downgrade system, like the system it replaces, is a little fragile. It trusts that the table in data/gamedata/downgrades.json is accurate and doesn't contain any missions with incompatible hardware requirements. If that assumption is mistaken, the results are unspecified.

The hardware requirements is a pretty strong requirement - for one, it essentially prevents any two-crew missions from being downgraded. Joint missions & single launch missions cannot be swapped. Don't downgrade anything which has an optional docking module to a mission which requires one - the hardware has already been assigned in the VAB. It's possible to downgrade a mission with a DM to one without - the DM will be assigned until the next turn, but probably won't be consumed; it may also be possible to free it up by reassigning hardware in the VAB.

I'll try to tighten up the hardware issues at some point and make the system more resilient. Currently, there isn't much protection surrounding those issues (Ex: #292) so I think the added functionality - used properly - is worth it.

**Comment by peyre** [comment]
Hm, I just scheduled a manned suborbital but it wouldn't let me downgrade it to unmanned.  It would only let me downgrade it to a "SCRUBBED MISSION".

![Scrubbed](https://user-images.githubusercontent.com/3118014/68073954-da1d2700-fd52-11e9-9f96-fcf4dbae8ae8.png)

**Comment by rnyoakum** [comment]
Did you add the option in the downgrades.json file? I don't know all the mission eccentricities, so I didn't add the unmanned missions as options.

**Comment by peyre** [comment]
Oh!  No, I didn't change any files; I just downloaded the latest copy of the repository and compiled.

**Comment by rnyoakum** [comment]
Yeah - I thought it might be nice to have as a feature. I bypassed the popup messages asking for confirmation which normally appear when scrubbing a mission because the UI didn't work well with them. If you don't like having the ability to scrub in Mission Control, or if you feel it shouldn't happen without the popup messages, I can disable them - after all, it's always possible to scrub via the Launch Pad.

**Comment by peyre** [comment]
It seems a little dangerous offhand, but that's probably just my natural hesitance about new ideas.  I don't object to making it possible to scrub from Mission Control.  If you like the idea, we should probably leave it in.  It does still let you scroll back to your original mission, which is the important thing to my mind.

**Comment by rnyoakum** [comment]
I have a few hesitations - when a player chooses to scrub in the VAB or the Launch Pad it brings up a dialogue asking them to confirm. I had to skip the warnings because I didn't have it working well with the UI, so I removed that aspect. I might disable scrubbing from Mission Control until I implement a warning system at least.

**Comment by peyre** [comment]
Oh, ok.  I don't feel strongly about it either way, so long as you can always scroll back to the original mission.  Personally I'd be more excited about warning dialogs at the end of your turn (#166).  That tripped me up again yesterday, grr.

**Comment by NooneAtAll3-2** [comment]
is this one still unimplemented?

**Comment by peyre** [comment]
166 is implemented, but the ability to downgrade to an unmanned mission hasn't been.  It would be handy - your capsule isn't ready after all (maybe you got bad R&D rolls, maybe you got a downgrade in the newscast), but it would be handy to send up the flight without people aboard to test the equipment.
